### PR TITLE
Continue bringing libpacemaker up to current guidelines

### PR DESCRIPTION
**Pull request by kgaillot**
This focuses on changes that will help address some colocation issues. The most significant part is moving some shared clone/bundle code to a new source file.

### CODE REVIEW

**Comment by kgaillot** [comment]
@nrwahl2 , can you review this when you get a chance? Low priority

File: lib/pacemaker/pcmk_sched_instances.c [hunk 1]
```diff
@@ -224,24 +224,41 @@ append_parent_colocation(pe_resource_t * rsc, pe_resource_t * child, gboolean al
...
+void
+distribute_children(pe_resource_t *rsc, GList *children, int max,
+                    int per_host_max, pe_working_set_t *data_set)
+{
+    // Reuse node count to track number of assigned instances
+    unsigned int available_nodes = reset_allowed_node_counts(rsc);
+
+    int loop_max = 0;
+    int allocated = 0;
+    bool all_coloc = false;
```
Comment (nrwahl2) [inline lib/pacemaker/pcmk_sched_instances.c:306]: Now we can assign the `all_coloc` value in the declaration line if you want

File: lib/pacemaker/pcmk_sched_instances.c [hunk 2]
```diff
@@ -249,6 +249,51 @@ reset_allowed_node_counts(pe_resource_t *rsc)
...
+ * \param[in] optimal_per_node  Optimal number of instances per node
+ *
+ * \return Instance's current node if still available, otherwise NULL
+ */
+static pe_node_t *
+preferred_node(const pe_resource_t *rsc, const pe_resource_t *instance,
+               int optimal_per_node)
+{
+    pe_node_t *node = NULL;
+    pe_node_t *parent_node = NULL;
```
Comment (nrwahl2) [inline lib/pacemaker/pcmk_sched_instances.c:267]: both const

**Review by nrwahl2 — COMMENTED** [review]
Excellent, slightly fewer headaches :)

File: lib/pengine/clone.c [hunk 3]
```diff
@@ -30,6 +30,23 @@
 #define UNPROMOTED_INSTANCES RSC_ROLE_UNPROMOTED_S
 #endif
 
+/*!
+ * \internal
+ * \brief Return the maximum number of clone instances allowed to be run
+ *
+ * \param[in] clone  Promotable clone or clone to check
```
[the remark this answers is not in the document]
Comment (kgaillot) [reply lib/pengine/clone.c:37]: Yep, that was the intention but I forgot :)

File: lib/pacemaker/pcmk_sched_instances.c [hunk 4]
```diff
@@ -0,0 +1,558 @@
...
+                             pcmk_is_set(action->flags, pe_action_runnable),
+                             pcmk_is_set(action->flags, pe_action_pseudo));
+                *starting = TRUE;
+            }
+        }
+    }
+}
+
+void
+clone_create_pseudo_actions(pe_resource_t *rsc, GList *children,
```
[the remark this answers is not in the document]
Comment (kgaillot) [reply lib/pacemaker/pcmk_sched_instances.c:436]: That's in a later batch

File: lib/pacemaker/pcmk_sched_instances.c [hunk 5]
```diff
@@ -18,58 +18,86 @@
...
+    pe_rsc_trace(instance, "%s can run on %s (%d already running)",
+                 instance->id, pe__node_name(node), allowed_node->count);
+    return true;
+}
+
+/*!
+ * \internal
+ * \brief Ban a clone instance from its allowed nodes that are unavailable
+ *
+ * \param[in,out] rsc           Clone instance
```
[the remark this answers is not in the document]
Comment (kgaillot) [reply lib/pacemaker/pcmk_sched_instances.c:81]: Whoops, yep

File: lib/pacemaker/pcmk_sched_instances.c [hunk 6]
```diff
@@ -18,58 +18,86 @@
...
+ * \param[in]     max_per_node  Maximum instances allowed to run on a node
+ */
+static void
+ban_unavailable_allowed_nodes(pe_resource_t *instance, int max_per_node)
+{
+    if (instance->allowed_nodes != NULL) {
+        GHashTableIter iter;
+        pe_node_t *allowed_node = NULL;
```
[the remark this answers is not in the document]
Comment (kgaillot) [reply lib/pacemaker/pcmk_sched_instances.c:89]: works

File: lib/pacemaker/pcmk_sched_instances.c [hunk 7]
```diff
@@ -18,58 +18,86 @@
...
-
-    if (local_node == NULL) {
+    allowed_node = pcmk__top_allowed_node(instance, node);
+    if (allowed_node == NULL) {
         crm_warn("%s cannot run on %s: node not allowed",
-                 rsc->id, pe__node_name(node));
-        goto bail;
-
-    } else if (local_node->weight < 0) {
-        common_update_score(rsc, node->details->id, local_node->weight);
```
[the remark this answers is not in the document]
Comment (kgaillot) [reply lib/pacemaker/pcmk_sched_instances.c]: What did the original code do after this?

File: lib/pacemaker/pcmk_sched_instances.c [hunk 8]
```diff
@@ -18,58 +18,86 @@
...
+        pe_rsc_trace(instance,
+                     "%s cannot run on %s because it cannot run resources",
+                     instance->id, pe__node_name(node));
+        return false;
     }
 
-    local_node = pcmk__top_allowed_node(rsc, node);
-
-    if (local_node == NULL) {
+    allowed_node = pcmk__top_allowed_node(instance, node);
```
[the remark this answers is not in the document]
Comment (kgaillot) [reply lib/pacemaker/pcmk_sched_instances.c:50]: Bundles are a bit odd; the replicas are their own structs but not in the resource hierarchy. A bundle replica's container has the bundle as the immediate parent.

File: lib/pacemaker/pcmk_sched_instances.c [hunk 9]
```diff
@@ -216,7 +211,10 @@ distribute_children(pe_resource_t *rsc, GList *children, GList *nodes,
         }
     }
 
-    all_coloc = (max < available_nodes) ? true : false;
+    /* Include positive colocation preferences of dependent resources
```
[the remark this answers is not in the document]
Comment (kgaillot) [reply lib/pacemaker/pcmk_sched_instances.c:308]: good point

File: lib/pacemaker/pcmk_sched_instances.c [hunk 10]
```diff
@@ -100,68 +100,100 @@ ban_unavailable_allowed_nodes(pe_resource_t *instance, int max_per_node)
...
+                 "Assigning %s (preferring %s, using %s parent colocations)",
+                 instance->id,
+                 ((prefer == NULL)? "no node" : prefer->details->uname),
+                 (all_coloc? "all" : "essential"));
+
+    if (!pcmk_is_set(instance->flags, pe_rsc_provisional)) {
+        // Instance is already assigned
+        return instance->fns->location(instance, NULL, FALSE) != NULL;
+
+    } else if (pcmk_is_set(instance->flags, pe_rsc_allocating)) {
```
[the remark this answers is not in the document]
Comment (kgaillot) [reply lib/pacemaker/pcmk_sched_instances.c:137]: why not

File: lib/pacemaker/pcmk_sched_instances.c [hunk 11]
```diff
@@ -100,68 +100,100 @@ ban_unavailable_allowed_nodes(pe_resource_t *instance, int max_per_node)
...
+        pe_rsc_debug(instance,
+                     "Dependency assignment loop detected involving %s",
```
[the remark this answers is not in the document]
Comment (kgaillot) [reply lib/pacemaker/pcmk_sched_instances.c:139]: Both are confusing :) I'll go with "Assignment loop detected involving %s colocations"

File: lib/pacemaker/pcmk_sched_instances.c [hunk 12]
```diff
@@ -100,68 +100,100 @@ ban_unavailable_allowed_nodes(pe_resource_t *instance, int max_per_node)
...
 
-static pe_node_t *
-allocate_instance(pe_resource_t *rsc, pe_node_t *prefer, gboolean all_coloc,
-                  int limit, pe_working_set_t *data_set)
+/*!
+ * \internal
+ * \brief Choose a node for an instance
+ *
+ * \param[in,out] instance      Clone instance or bundle replica container
+ * \param[in]     prefer        If not NULL, node to prefer for early assignment
```
[the remark this answers is not in the document]
Comment (kgaillot) [reply lib/pacemaker/pcmk_sched_instances.c:108]: good point

File: lib/pacemaker/pcmk_sched_instances.c (hunk 2 above)
Comment (kgaillot) [reply lib/pacemaker/pcmk_sched_instances.c:267]: good point

**Comment by kgaillot** [comment]
Added 3 commits on top to address review

File: lib/pacemaker/pcmk_sched_instances.c [hunk 13]
```diff
@@ -100,68 +100,100 @@ ban_unavailable_allowed_nodes(pe_resource_t *instance, int max_per_node)
...
+ * \param[in]     all_coloc     If true (indicating that there are more
```
[the remark this answers is not in the document]
Comment (kgaillot) [reply lib/pacemaker/pcmk_sched_instances.c:110]: I think the logs are easier to follow as is

File: lib/pacemaker/pcmk_sched_instances.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (nrwahl2) [reply lib/pacemaker/pcmk_sched_instances.c]: Ahh, I was thinking we returned somewhere before the bail label, or else we would have just said `goto bail` here. Wrong :)

File: lib/pacemaker/pcmk_sched_instances.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (nrwahl2) [reply lib/pacemaker/pcmk_sched_instances.c:50]: True for replica containers. A bundle also has clone instance grandchildren IIRC. Probably fine (and I think those instances run on the replica containers IIRC, so top-level parent allowed nodes wouldn't be relevant) -- didn't dig into that.

File: lib/pacemaker/pcmk_sched_instances.c (hunk 13 above)
[the remark this answers is not in the document]
Comment (nrwahl2) [reply lib/pacemaker/pcmk_sched_instances.c:110]: works for me

File: lib/pacemaker/pcmk_sched_instances.c (hunk 11 above)
[the remark this answers is not in the document]
Comment (nrwahl2) [reply lib/pacemaker/pcmk_sched_instances.c]: Sounds good

**Review by nrwahl2 — APPROVED** [review]

**Pull request merged**